File: TestExecution/TestRuntime/Logger.cs
using Microsoft.Diagnostics.Tracing;

namespace R4nd0mApps.TddStud10.TestRuntime.Diagnostics
{
    [EventSource(Name = Constants.EtwProviderNameTestRuntime)]
    internal sealed class Logger : EventSource
    {
        public static Logger I = new Logger();

        [Event(1, Level = EventLevel.Informational)]
        public void Log(string message)
        {
            WriteEvent(1, message);
        }

        [Event(2, Level = EventLevel.Error)]
        internal void LogError(string message)
        {
            base.WriteEvent(2, message);
        }

        [NonEvent]
        public void LogInfo(string format, params object[] args)
        {
            if (IsEnabled())
            {
                Log(string.Format(format, args));
            }
        }

        [NonEvent]
        public void LogError(string format, params object[] args)
        {
            if (IsEnabled(EventLevel.Error, EventKeywords.All))
            {
                LogError(string.Format(format, args));
            }
        }
    }
}

File: Engine.Core/Logger.fs
namespace R4nd0mApps.TddStud10.Engine.Diagnostics

open System
open Microsoft.Diagnostics.Tracing
open R4nd0mApps.TddStud10

[<Sealed>]
[<EventSource(Name = Constants.EtwProviderNameEngine)>]
type internal Logger() = 
    inherit EventSource()
    static let i = new Logger()
    static member I = i
    
    [<Event(1, Level = EventLevel.Informational)>]
    member public __.LogInfo(message : string) = base.WriteEvent(1, message)
    
    [<Event(2, Level = EventLevel.Error)>]
    member public __.LogError(message : string) = base.WriteEvent(2, message)
    
    [<NonEvent>]
    member public this.LogInfo(format : string, [<ParamArray>] args : Object array) = 
        if (base.IsEnabled(EventLevel.Informational, EventKeywords.All)) then this.LogInfo(String.Format(format, args))
    
    [<NonEvent>]
    member public this.LogError(format : string, [<ParamArray>] args : Object array) = 
        if (base.IsEnabled(EventLevel.Error, EventKeywords.All)) then this.LogError(String.Format(format, args))
    
    [<NonEvent>]
    static member private invokeLogf level logFn = 
        let l = 
            if Logger.I.IsEnabled(level, EventKeywords.All) then logFn
            else ignore
        Printf.ksprintf l
    
    [<NonEvent>]
    static member public logInfof format = Logger.invokeLogf EventLevel.Informational Logger.I.LogInfo format
    
    [<NonEvent>]
    static member public logErrorf format = Logger.invokeLogf EventLevel.Error Logger.I.LogError format
